### PR TITLE
Introduce `binary` and `binary_basename` in GSC templates

### LINKED ISSUE #65: Failure in gsc sign-image due to protobuf major upgrade
This appears to be a failure due to update in major release version of protobuf from 3.20.1 to 4.21.0 - https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates

Below are the logs when I attempt to run `gsc sign-image IMAGE_NAME SIGNING_KEY_FILE`

```
...
Step 2/6 : COPY gsc-signer-key.pem /gsc-signer-key.pem

 ---> aa1594fff42b
Step 3/6 : RUN export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib -type d -path '*/site-packages')" && gramine-sgx-sign       --key /gsc-signer-key.pem       --manifest /entrypoint.manifest       --output /entrypoint.manifest.sgx

 ---> Running in a3c8fd7333a5
Traceback (most recent call last):
  File "/gramine/meson_build_output/bin/gramine-sgx-sign", line 11, in <module>
    from graminelibos import (
  File "/gramine/meson_build_output/lib/python3.8/site-packages/graminelibos/__init__.py", line 24, in <module>
    from .sgx_get_token import get_token
  File "/gramine/meson_build_output/lib/python3.8/site-packages/graminelibos/sgx_get_token.py", line 9, in <module>
    from . import _aesm_pb2 as aesm_pb2
  File "/gramine/meson_build_output/lib/python3.8/site-packages/graminelibos/_aesm_pb2.py", line 38, in <module>
    _descriptor.FieldDescriptor(
  File "/usr/local/lib/python3.8/dist-packages/google/protobuf/descriptor.py", line 560, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors cannot not be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

More information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates

Removing intermediate container a3c8fd7333a5
```

### PR DESCRIPTION
**Pull request by dimakuv**
## Description of the changes 

This PR contains ~three~ ~two~ one commit (~rather independent but hit these three issues while trying OpenVINO~):

~- Set `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python` during GSC steps. Core Gramine uses a pre-generated `_pb2.py` file, which became out of date with the new release of the Google Protobuf Python lib. GSC started to failing because of this during the build & sign steps. Technically, core Gramine should not use the pre-generated file but instead create this file on demand. But this requires some modifications in core Gramine. A simple workaround for this issue is suggested by Protobuf devs: set `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python`.~

- Previously, GSC templates used `which {{binary}}` to find the absolute path to the entrypoint binary. GSC sets the `binary` variable here to the basename of the Docker image's ENTRYPOINT. So the `which {{binary}}` command worked well for binaries found in PATH, but failed for binaries not under PATH but rather specified using absolute/relative path. This commit changes GSC to set the `binary` variable to the full path. For the cases when the basename of the binary is needed, we introduce a new variable `binary_basename`.

~- [Examples] Fix OpenVINO to build and run. OpenVINO native build got affected by Google Protobuf Python lib update and requires the `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python` workaround. Also, OpenVINO now has a proper ENTRYPOINT, with an absolute path to the executable (previously it was a relative path to the symlink to the executable, which confused Gramine and led to failures).~

~Fixes #65.~

## How to test this PR? 

- To test PATH binary, try the classic Python example (https://gramine.readthedocs.io/projects/gsc/en/latest/#example).

- To test non-PATH absolute-path binary, try our OpenVINO example.

- To test non-PATH relative-path binary, try the modified `ubuntu18.04-bash` example from `tests/`. Modify the Dockerfile like this:
```diff
diff --git a/test/ubuntu18.04-bash.dockerfile b/test/ubuntu18.04-bash.dockerfile
@@ -2,4 +2,6 @@ From ubuntu:18.04

 RUN apt-get update

-CMD ["bash"]
+WORKDIR /bin/
+
+ENTRYPOINT ["./bash"]
```

- To test non-PATH relative-path binary (another example), try the modified `ubuntu18.04-bash` example from `tests/`. Modify the Dockerfile like this:
```diff
diff --git a/test/ubuntu18.04-bash.dockerfile b/test/ubuntu18.04-bash.dockerfile
@@ -2,4 +2,4 @@ From ubuntu:18.04

 RUN apt-get update

-CMD ["bash"]
+ENTRYPOINT ["bin/bash"]

```

---
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)

### CODE REVIEW

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: 0 of 6 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel)

*[a discussion](https://reviewable.io/reviews/gramineproject/gsc/71#-N5ZmySk11SW5zI4SQVA:-N5ZmySk11SW5zI4SQVB:b-c2vmir) (no related file):*
@veenasai2 and/or @aneessahib, could you take a look at this PR? Do the changes sound reasonable to you?

___

**Review by rranjan3 — DISMISSED** [review]
Able to verify the 20.04 distro

**Review by aneessahib — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: 0 of 6 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv)

___
*[`templates/Dockerfile.common.build.template` line 46 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5czFIq7zYUOWRQK2pz:-N5czFIq7zYUOWRQK2q-:b-5etp86) ([raw file](https://github.com/gramineproject/gsc/blob/1659f30a5362516c44e460107303e15014cc6e28/templates/Dockerfile.common.build.template#L46)):*
> ```
> # Docker entrypoint/cmd typically contains only the basename of the executable so create a symlink
> RUN cd /gramine/app_files/ \
>     && which {{binary_full_path}} | xargs ln -s || true
> ```

Hi @dimakuv  - Quick clarification here, probably something trivial. Why would we do `which` on a full binary path? This wouldn't return anything useful I think. And the whole symbolic link generation too might not happen.

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: 0 of 6 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @aneessahib)

___
*[`templates/Dockerfile.common.build.template` line 46 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5czFIq7zYUOWRQK2pz:-N5d2L0D6lVKGgHNzXIL:b-3a5h5) ([raw file](https://github.com/gramineproject/gsc/blob/1659f30a5362516c44e460107303e15014cc6e28/templates/Dockerfile.common.build.template#L46)):*
<details><summary><i>Previously, aneessahib (Anees Sahib) wrote…</i></summary><blockquote>

Hi @dimakuv  - Quick clarification here, probably something trivial. Why would we do `which` on a full binary path? This wouldn't return anything useful I think. And the whole symbolic link generation too might not happen.
</blockquote></details>

`which` correctly returns the full binary path if given the full binary path. Random example:
```
$ which ab
/usr/bin/ab

$ which /usr/bin/ab
/usr/bin/ab
```

So `which` works correctly in two cases:
1. The "short" binary name (in the example above, simply `which ab`)
2. The full binary name

Note that the first case works *only* if the binary can be found under one of the `PATH` paths. This is not the case with e.g. our OpenVINO example. So that's why I'm using the absolute path in that case (see my other changes in this PR).

And after `which` successfully returns `/usr/bin/ab`, then the symbolic link is correctly generated:
```
/gramine/app_files/ab  -->  /usr/bin/ab
```

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: 0 of 6 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @aneessahib)

___
*[`templates/Dockerfile.common.build.template` line 46 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5czFIq7zYUOWRQK2pz:-N5dHzrR4tSsWVwHXKFe:b-xlu4g7) ([raw file](https://github.com/gramineproject/gsc/blob/1659f30a5362516c44e460107303e15014cc6e28/templates/Dockerfile.common.build.template#L46)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`which` correctly returns the full binary path if given the full binary path. Random example:
```
$ which ab
/usr/bin/ab

$ which /usr/bin/ab
/usr/bin/ab
```

So `which` works correctly in two cases:
1. The "short" binary name (in the example above, simply `which ab`)
2. The full binary name

Note that the first case works *only* if the binary can be found under one of the `PATH` paths. This is not the case with e.g. our OpenVINO example. So that's why I'm using the absolute path in that case (see my other changes in this PR).

And after `which` successfully returns `/usr/bin/ab`, then the symbolic link is correctly generated:
```
/gramine/app_files/ab  -->  /usr/bin/ab
```
</blockquote></details>

Ok, I was wrong, Anees is right. `which` only outputs the full name when it finds the application under one of the `PATH` items.

Quick negative test:
```
~/gramineproject/gramine/CI-Examples/helloworld$ which /home/dimakuv/gramineproject/gramine/CI-Examples/helloworld

~/gramineproject/gramine/CI-Examples/helloworld$ echo $?
1
```

I need to fix my PR.

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: 0 of 6 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @aneessahib)

___
*[`templates/Dockerfile.common.build.template` line 46 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5czFIq7zYUOWRQK2pz:-N5dJtxME24zw6ToU4fE:b-6og5vl) ([raw file](https://github.com/gramineproject/gsc/blob/1659f30a5362516c44e460107303e15014cc6e28/templates/Dockerfile.common.build.template#L46)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Ok, I was wrong, Anees is right. `which` only outputs the full name when it finds the application under one of the `PATH` items.

Quick negative test:
```
~/gramineproject/gramine/CI-Examples/helloworld$ which /home/dimakuv/gramineproject/gramine/CI-Examples/helloworld

~/gramineproject/gramine/CI-Examples/helloworld$ echo $?
1
```

I need to fix my PR.
</blockquote></details>

Ah, no, my logic is correct. Me and Anees were trying `which /home/dimakuv/gramineproject/gramine/CI-Examples/helloworld` -- this is a path, not a full-named executable.

This corrected example works:
```
$ which /home/dimakuv/gramineproject/gramine/CI-Examples/helloworld/helloworld
/home/dimakuv/gramineproject/gramine/CI-Examples/helloworld/helloworld
```

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @aneessahib and @dimakuv)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5db7cE-tpIwXTks1Qn):*
I would also explain what this workaround actually does: uses python implementation for protobuf.
___
*[`-- commits` line 22 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5dbMjP-PoaB5lXC-xx):*
After reading this commit message I have no idea what's going on here. From a quick glance at the rest: the problem was that `which` couldn't find the binary in PATH? I would rephrase this to describe what the actual problem was (right now it says that there were errors, period).
___
*[`gsc.py` line 63 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5dcsbL7Da_3roSTjFa:-N5dcsbL7Da_3roSTjFb:bmr76ia) ([raw file](https://github.com/gramineproject/gsc/blob/1659f30a5362516c44e460107303e15014cc6e28/gsc.py#L63)):*
> ```Python
> 
>     # Set binary to first executable in entrypoint
>     binary_full_path = entrypoint[0]
> ```

Hm, this still could be `./something/app.exe` if that's what somebody has in their dockerfile, i.e. this is not necessarily a full path

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @aneessahib, @boryspoplawski, and @dimakuv)

___
*[`gsc.py` line 63 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5dcsbL7Da_3roSTjFa:-N5eDBWK1afqJ1BqQpwG:b-89b4b9) ([raw file](https://github.com/gramineproject/gsc/blob/1659f30a5362516c44e460107303e15014cc6e28/gsc.py#L63)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Done. The documentation on Docker variables (like a format of `Entrypoint`) is quite vague, so I added ad-hoc logic. Should be good enough.
</blockquote></details>

I think this is still wrong. `work_dir == "/root"` and `entrypoint[0] == "some_dir/my_app"` sounds like a viable option.
___
*[`gsc.py` line 50 at r2](https://reviewable.io/reviews/gramineproject/gsc/71#-N5eDomV2oCGwwQ7BQeK:-N5eDomV2oCGwwQ7BQeL:b-k0gje6) ([raw file](https://github.com/gramineproject/gsc/blob/4bba671b535173864ea402730f2b0da5a88eb98b/gsc.py#L50)):*
> ```Python
> 
> 
> def extract_binary_cmd_workdir_from_image_config(config, env):
> ```

I think the original name was supposed to mean "extract binary's commandline", now it's just a (random) concatenation of words

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @aneessahib, @boryspoplawski, and @dimakuv)

___
*[`gsc.py` line 63 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5dcsbL7Da_3roSTjFa:-N5ez_vxA40P5ODdkJz3:b-5d0ej8) ([raw file](https://github.com/gramineproject/gsc/blob/1659f30a5362516c44e460107303e15014cc6e28/gsc.py#L63)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Yeah. Ok, more hackery added.
</blockquote></details>

Ok, this is hacky, but should work. The only issue left is that the variable name is wrong - it might as well be just the name of the binary to be found in `PATH` (i.e. not full path at all).

**Review by boryspoplawski — DISMISSED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners (waiting on @aneessahib and @boryspoplawski)

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners (waiting on @aneessahib and @boryspoplawski)

___
*[`gsc.py` line 63 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5dcsbL7Da_3roSTjFa:-N5fR3W50mt08_9NOYvt:b-7r20vm) ([raw file](https://github.com/gramineproject/gsc/blob/1659f30a5362516c44e460107303e15014cc6e28/gsc.py#L63)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Ok, this is hacky, but should work. The only issue left is that the variable name is wrong - it might as well be just the name of the binary to be found in `PATH` (i.e. not full path at all).
</blockquote></details>

Done (looks like you approved before I put a comment :)). I swapped `binary_full_path` and `binary` to `binary` and `binary_basename`. I think the naming is more logical now.

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @aneessahib, @boryspoplawski, and @dimakuv)

___
*[`-- commits` line 22 at r1](https://reviewable.io/reviews/gramineproject/gsc/71#-N5dbMjP-PoaB5lXC-xx):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Not done?
</blockquote></details>

You're too fast :)

I'm still testing this PR, will update everything shortly.

**Review by boryspoplawski — DISMISSED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: all files reviewed, 1 unresolved discussion (waiting on @aneessahib)

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: all files reviewed, 1 unresolved discussion (waiting on @aneessahib)

*[a discussion](https://reviewable.io/reviews/gramineproject/gsc/71#-N6O2rV53Kqnhu-ybD73:-N6O2rV53Kqnhu-ybD74:b-ksk0p8) (no related file):*
I verified again, everything works. (For OpenVINO to work, see also https://github.com/gramineproject/gsc/pull/73)

___

**Review by dimakuv — COMMENTED** [review]
> Dismissed @aneessahib from a discussion.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/71)* status: :shipit: complete! all files reviewed, all discussions resolved
